### DISCUSSION TITLE
Practice data

### DISCUSSION THREAD

**Discussion by Majd-ui** [theOehrly/Fast-F1#118] [Q&A]
How long till I can access the practice data

**Comment by walteij** [comment]
Wel, I tried getting the practice data using the livetiming, but am getting an error when trying to load it:

`Traceback (most recent call last):
  File "%APPDATA%\Python\Python39\site-packages\fastf1\core.py", line 1407, in _drivers_results_from_ergast
    data = ergast.fetch_results(
  File "%APPDATA%\Python\Python39\site-packages\fastf1\ergast.py", line 25, in fetch_results
    return _parse_ergast(fetch_day(year, gp, day))[0][sel]
IndexError: list index out of range
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "%APPDATA%\Python\Python39\site-packages\fastf1\core.py", line 1056, in load
    raise NoLapDataError
fastf1.core.NoLapDataError: Failed to load session because the API did not provide any usable data.`

**Reply by theOehrly** [reply]
That looks like a bug. It shouldn't throw API related errors if you provide a recording of the livetiming data. Potentially I messed something up there, as the code that is throwing the error is new in v2.2

**Reply by walteij** [reply]
Could also be that I messed up something in my livetiming recording. For some strange reason my NIC decided to stop working for 20 seconds (which was at the end of the red flag caused by Ocons failing sidepod).

But if you want me to test something, let me know.

**Reply by theOehrly** [reply]
I'll test it with a recording of FP2.

**Reply by markste-in** [reply]
Have the same issue

```
sess = ff1.get_session(2022,1,'FP1')
laps = sess.load()
```

```
core           INFO 	Loading data for Bahrain Grand Prix - Practice 1 [v2.2.0]
api            INFO 	Using cached data for driver_info
core          ERROR 	Failed to load data from Ergast API!
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniforge/base/envs/f1alpha/lib/python3.10/site-packages/fastf1/core.py", line 1407, in _drivers_results_from_ergast
    data = ergast.fetch_results(
  File "/opt/homebrew/Caskroom/miniforge/base/envs/f1alpha/lib/python3.10/site-packages/fastf1/ergast.py", line 25, in fetch_results
    return _parse_ergast(fetch_day(year, gp, day))[0][sel]
IndexError: list index out of range
```

**Reply by theOehrly** [reply]
``core          ERROR 	Failed to load data from Ergast API!``
This line is valid. It takes longer until the additional information on Ergast is available after a session than it takes for the timing and telemetry data to be available. The following traceback shouldn't be printed though because the error really is expected here. But all the data except for the session results should still be loaded correctly, and you can work with it. 
So this is not really a problem. It looks worse than it is.

**Reply by markste-in** [reply]
Unfortunately I just get NoneType back.

It also tells me that I cannot use `load() `with `with_telemetry=True`

```
sess = ff1.get_session(2022,1,'FP1')
laps = sess.load(with_telemetry=True)
```
`TypeError: Session.load() got an unexpected keyword argument 'with_telemetry'`

`load_laps() `works even tho it tells me it is depreceated

**Reply by theOehrly** [reply]
Yes, `session.load` does not work exactly the same as `.load_laps` does.
Because `.load` really does not only load lap timing data but also telemetry, driver lists, results and weather data, I decided that it does not make sense, in terms of api consistency, to then only return the laps data. Therefore, after calling `.load`, the data can be accessed (as was previously already the case) through the properties `session.laps`, `session.results`, `session.weather_data` and others.
Also, `.load` now has more keyword arguments to allow finer control over what data should be loaded. These are named differently and all of them default to true.
`.load_laps` in the latests release only wraps `.load` to provide backwards compatibility.

**Comment by theOehrly** [comment]
I'd say it takes 30 to 60 minutes usually

**Reply by Majd-ui** [reply]
Ok thank you

**Reply by walteij** [reply]
Data is there :)
![weather_Bahrein_2022_FP1 ](https://user-images.githubusercontent.com/64311753/159014362-18795e80-c7ff-4a88-83b2-1b1d13a4459f.png)
![2022_Bahrein_FP1_GAS_fastestlap](https://user-images.githubusercontent.com/64311753/159014382-d74fc4f3-1a13-49ff-ba66-a9bd9486edec.png)
![strategy_2022_Bahrein_FP1](https://user-images.githubusercontent.com/64311753/159014399-5a001abe-5841-4037-98fc-31d017264a89.png)

**Reply by Majd-ui** [reply]
yes it is

**Comment by markste-in** [comment]
Just a simple question ( prob. just doing it wrong): There seems to be no speed / time data or does it take a little longer to process it?

```
sess = ff1.get_session(2022,1,'FP1')
laps = sess.load()
for d in sess.results.Abbreviation: # or sess.driver
    try:
        sess.laps.pick_driver(d).get_telemetry()
    except Exception as e:
        print(f'{d} : {e}')
```

```
VER : 'SessionTime'
SAI : Telemetry does not contain required channels 'Time' and 'Speed'.
MSC : Telemetry does not contain required channels 'Time' and 'Speed'.
HAM : Telemetry does not contain required channels 'Time' and 'Speed'.
OCO : Telemetry does not contain required channels 'Time' and 'Speed'.
HUL : Telemetry does not contain required channels 'Time' and 'Speed'.
ZHO : Telemetry does not contain required channels 'Time' and 'Speed'.
ALB : 'SessionTime'
TSU : Telemetry does not contain required channels 'Time' and 'Speed'.
MAG : Telemetry does not contain required channels 'Time' and 'Speed'.
STR : Telemetry does not contain required channels 'Time' and 'Speed'.
LEC : Telemetry does not contain required channels 'Time' and 'Speed'.
PER : 'SessionTime'
GAS : 'SessionTime'
LAT : Telemetry does not contain required channels 'Time' and 'Speed'.
NOR : Telemetry does not contain required channels 'Time' and 'Speed'.
RIC : Telemetry does not contain required channels 'Time' and 'Speed'.
RUS : Telemetry does not contain required channels 'Time' and 'Speed'.
BOT : Telemetry does not contain required channels 'Time' and 'Speed'.
```

**Reply by markste-in** [reply]
oh nevermind... it is `get_car_data()` now. Didn't know that changed

**Reply by theOehrly** [reply]
It hasn't changed. `get_telemetry` should work too. Both have existed since quite some time and are slightly different.

**Reply by markste-in** [reply]
hmmm... i can't get it to work anymore with `get_telemetry`. Same code works with `get_car_data` works 🤔

**Reply by theOehrly** [reply]
That's weird. I'll have to test it and see why that is the case. 
But, if you're code works with `get_car_data`, you should prefer that over `get_telemetry`. You will get somewhat better results, because no interpolation of the source data is applied.

**Reply by theOehrly** [reply]
@markste-in Ok, so this is a bug that occurs when getting telemetry for multiple laps but only in some cases where NaT values exist in the laps data. This bug has to have existed since quite a while. It will be fixed in the next release.

**Reply by markste-in** [reply]
Yes I used a workaround and did a  try/except with a `if "Speed" in df.columns`
